swim/heal_via_discover_provider.go

```diff
@@ -0,0 +1,190 @@
+		started:          make(chan struct{}, 1),
+		quit:             make(chan struct{}),
+	}
+}
+
+// Start the partition healing loop
+func (h *discoverProviderHealer) Start() {
+	// check if started channel is already filled
+	// if not, we start a new loop
+	select {
+	case h.started <- struct{}{}:
+	default:
+		return
+	}
+
+	go func() {
+		for {
+			// attempt heal with the pro
+			if rand.Float64() < h.Probability() {
+				h.Heal()
+			}
+
+			// loop or quit
+			select {
+			case <-h.node.clock.After(h.period):
+			case <-h.quit:
+				return
+			}
+		}
+	}()
+}
+
+// Stop the partition healing loop.
+func (h *discoverProviderHealer) Stop() {
+	// if started, consume and send quit signal
+	// if not started this is noop
+	select {
+	case <-h.started:
+		h.quit <- struct{}{}
+	default:
+	}
+}
+
+// Probability returns the probability when a heal should be attempted
+// we want to throttle the heal attempts to alleviate pressure on the
+// discover provider.
+func (h *discoverProviderHealer) Probability() float64 {
+	// avoid division by zero.
+	if h.previousHostListSize < h.node.CountReachableMembers() {
+		h.previousHostListSize = h.node.CountReachableMembers()
+	}
+	if h.previousHostListSize < 1 {
+		h.previousHostListSize = 1
+	}
+	return h.baseProbabillity / float64(h.previousHostListSize)
+}
+
+// Heal iterates over the hostList that the discoverProvider provides. If the
+// node encounters a host that is faulty or not in the membership, we pick that
+// node as a target to perform a partition heal with.
+//
+// If heal was attempted, returns identities of the target nodes.
+func (h *discoverProviderHealer) Heal() ([]string, error) {
+	h.node.emit(DiscoHealEvent{})
+	// get list from discovery provider
+	if h.node.discoverProvider == nil {
+		return []string{}, errors.New("discoverProvider not available to healer")
+	}
+
+	hostList, err := h.node.discoverProvider.Hosts()
+	if err != nil {
+		h.logger.Warn("healer unable to receive host list from discover provider")
+		return []string{}, err
+	}
+
+	h.previousHostListSize = len(hostList)
+	util.ShuffleStringsInPlace(hostList)
+
+	// collect the targets this node might want to heal with
+	var targets []string
+	for _, address := range hostList {
+		m, ok := h.node.memberlist.Member(address)
+		if !ok || statePrecedence(m.Status) >= statePrecedence(Faulty) {
+			targets = append(targets, address)
+		}
+	}
+
+	// filter hosts that we already know about and attempt to heal nodes that
+	// are complementary to the membership of this node.
+	var ret []string
+	failures := 0
+	maxFailures := 10
+	for len(targets) != 0 && failures < maxFailures {
+		target := targets[0]
+		targets = del(targets, target)
+
+		// try to heal partition
+		hostsOnOtherSide, err := AttemptHeal(h.node, target)
+
+		if err != nil {
+			h.logger.WithFields(log.Fields{
+				"error":   err.Error(),
+				"failure": failures,
+			}).Warn("heal attempt failed (10 in total)")
```

Comment:
I don't think we need the `(10 in total)` part of this log line. We log the `healer reached max failures` if we exceed the allowed number of failures.